File: hashing/counting/c++/max_number_target_word/max_number_target_word.cpp
// max_number_target_word.cpp
#include "max_number_target_word.hpp"
#include <string>
#include <unordered_map>
#include <limits.h>
#include <algorithm>

int MaxNumberTargetWord::loop(const std::string &text, const std::string &targetWord)
{
    std::unordered_map<char, int> textCounts;
    std::unordered_map<char, int> targetCounts;
    for (const auto &character : text)
    {
        auto result = textCounts.emplace(character, 1);
        // If key already exists, increment it
        if (!result.second)
        {
            // Access the second element of (character, count)
            result.first->second++;
        }
    }
    for (const auto &character : targetWord)
    {
        auto result = targetCounts.emplace(character, 1);
        // If key already exists, increment it
        if (!result.second)
        {
            // Access the second element of (character, count)
            result.first->second++;
        }
    }

    // If there are not enough unique characters, return 0
    if (textCounts.size() < targetCounts.size())
    {
        return 0;
    }

    // Iterate through each character in the target word, compute number of times we can satisfy the minimum number of each character needed to form the target word
    // E.g. target_counts = {'a': 2} (two letter 'a' required) and char_counts = {'a': 6}, then 6 // 2 = 3, then we can potentially form up to three instances of target word
    int maxNumberInstances = INT_MAX;
    for (const auto &character : targetWord)
    {
        // If not found in text, return 0 (we could also use unorderd_map::contains c++20)
        if (textCounts.find(character) == textCounts.end())
        {
            return 0;
        }
        int numTimeSatisfied = textCounts[character] / targetCounts[character];
        // Keep updating the minimum, i.e., we can't form more instances than any character's maximum potential allows
        maxNumberInstances = std::min(maxNumberInstances, numTimeSatisfied);
    }

    return maxNumberInstances;
}

int MaxNumberTargetWord::transformReduce(const std::string &text, const std::string &targetWord)
{
    std::unordered_map<char, int> textCounts;
    std::unordered_map<char, int> targetCounts;
    for (const auto &character : text)
    {
        auto result = textCounts.emplace(character, 1);
        // If key already exists, increment it
        if (!result.second)
        {
            // Access the second element of (character, count)
            result.first->second++;
        }
    }
    for (const auto &character : targetWord)
    {
        auto result = targetCounts.emplace(character, 1);
        // If key already exists, increment it
        if (!result.second)
        {
            // Access the second element of (character, count)
            result.first->second++;
        }
    }

    // If there are not enough unique characters, return 0
    if (textCounts.size() < targetCounts.size())
    {
        return 0;
    }

    // Iterate through each character in the target word, compute number of times we can satisfy the minimum number of each character needed to form the target word
    // E.g. target_counts = {'a': 2} (two letter 'a' required) and char_counts = {'a': 6}, then 6 // 2 = 3, then we can potentially form up to three instances of target word
    int maxNumberInstances = std::transform_reduce(
        targetCounts.begin(),
        targetCounts.end(),
        // Initial values is the maximum integer
        INT_MAX,
        // Reduce function applied to each consecutive output of the transform function below
        [](int currentMax, int nextMax)
        { return std::min(currentMax, nextMax); },
        // Transform function applied to each (character, counts)
        [&textCounts](const auto &targetCount)
        {
            const auto &[character, counts] = targetCount;
            // If not found, return 0
            if (textCounts.find(character) == textCounts.end())
            {
                return 0;
            }
            return textCounts[character] / counts;
        });

    return maxNumberInstances;
}

File: hashing/counting/python3/max_number_target_word.py
from collections import Counter
from sys import maxsize


def max_number_target_word(text: str, target_word: str) -> int:
    """
    Given a string `text`, use the characters of `text` to form as many instances of the word `target_word` as possible.

    Note: Use each character in text at most once.

    Parameters
    ----------
    text : str
        A string containing lower case english letters
    target_word : str
        A target word

    Returns
    -------
    int
        The maximum number of instances of the word `target_word` that can be formed
    """
    # O(n + m) to build
    char_counts = Counter(text)
    target_counts = Counter(target_word)

    # If text has fewer unique characters than target_word
    if len(char_counts) < len(char_counts):
        return 0

    max_num_instances = maxsize
    # Iterate through each character in the target word, compute number of times we can satisfy the minimum number of each character needed to form the target word
    # E.g. target_counts = {'a': 2} (two letter 'a' required) and char_counts = {'a': 6}, then 6 // 2 = 3, then we can potentially form up to three instances of target word
    for char in target_counts:
        # If a key is not found, 0 is the default, which means that no instance of the word can be formed
        if char_counts[char] == 0:
            return 0
        num_time_satisfied = char_counts[char] // target_counts[char]
        # Keep updating the minimum, i.e., we can't form more instances than any character's maximum potential allows
        max_num_instances = min(max_num_instances, num_time_satisfied)

    return max_num_instances


def main() -> int:
    texts = ["nlaebolko", "loonbalxballpoon", "leetcode", "balon"]
    for text in texts:
        max_num = max_number_target_word(text, "balloon")
        starting_message = f"Given the input {text}"
        ending_message = "the word 'balloon' from its characters"
        if max_num:
            print(
                f"{starting_message}, there can at most {max_num} instances of {ending_message}"
            )
        else:
            print(
                f"{starting_message}, it is not possible to form the {ending_message}"
            )

    return 0


if __name__ == "__main__":
    main()
